xcm/xcm-emulator/src/lib.rs

```diff
@@ -1048,42 +1117,109 @@ macro_rules! decl_test_bridges {
+	($chain:ident) => {
+		impl<Origin, Destination, Hops, Args>
+			$crate::CheckAssertion<Origin, Destination, Hops, Args> for $chain
+		where
+			Origin: Chain + Clone,
+			Destination: Chain + Clone,
+			Origin::RuntimeOrigin: $crate::OriginTrait<AccountId = $crate::AccountId32> + Clone,
+			Destination::RuntimeOrigin:
+				$crate::OriginTrait<AccountId = $crate::AccountId32> + Clone,
+			Hops: Clone,
+			Args: Clone,
+		{
+			fn check_assertion(test: $crate::Test<Origin, Destination, Hops, Args>) {
+				let chain_name = std::any::type_name::<$chain>();
+
+				<$chain>::execute_with(|| {
+					if let Some(dispatchable) = test.hops_dispatchable.get(chain_name) {
+						$crate::assert_ok!(dispatchable(test.clone()));
+					}
+					if let Some(assertion) = test.hops_assertion.get(chain_name) {
+						assertion(test);
+					}
+				});
+			}
+		}
+	};
+}
+
 #[macro_export]
 macro_rules! assert_expected_events {
 	( $chain:ident, vec![$( $event_pat:pat => { $($attr:ident : $condition:expr, )* }, )*] ) => {
 		let mut message: Vec<String> = Vec::new();
+		let mut events = <$chain>::events();
+
 		$(
+			let mut event_received = false;
 			let mut meet_conditions = true;
+			let mut index_match = 0;
 			let mut event_message: Vec<String> = Vec::new();
 
-			let event_received = <$chain>::events().iter().any(|event| {
-				$crate::log::debug!(target: concat!("events::", stringify!($chain)), "{:?}", event);
-
+			for (index, event) in events.iter().enumerate() {
+				// Have to reset the variable to override a previous partial match
+				meet_conditions = true;
 				match event {
 					$event_pat => {
+						event_received = true;
+						let mut conditions_message: Vec<String> = Vec::new();
+
 						$(
-							if !$condition {
-								event_message.push(format!(" - The attribute {:?} = {:?} did not met the condition {:?}\n", stringify!($attr), $attr, stringify!($condition)));
-								meet_conditions &= $condition
+							// We only want to record condition error messages in case it did not happened before
+							// Only the first partial match is recorded
+							if !$condition && event_message.is_empty() {
+								conditions_message.push(
+									format!(
+										" - The attribute {:?} = {:?} did not met the condition {:?}\n",
+										stringify!($attr),
+										$attr,
+										stringify!($condition)
+									)
+								);
 							}
+							meet_conditions &= $condition;
 						)*
-						true
+
+						// Set the index where we found a perfect match
+						if event_received && meet_conditions {
+							index_match = index;
+							break;
+						} else {
+							event_message.extend(conditions_message);
+						}
 					},
-					_ => false
+					_ => {}
 				}
-			});
+			}
 
 			if event_received && !meet_conditions  {
-				message.push(format!("\n\nEvent \x1b[31m{}\x1b[0m was received but some of its attributes did not meet the conditions:\n{}", stringify!($event_pat), event_message.concat()));
+				message.push(
```

Comment:
@NachoPal 
can we add all  `events` to the all messages?
in case, anything fails, we could see from the log all events, it could help to better analyse :)

Reply:
There is a log for that. If you want to debug you just need to run your tests with:
```
RUST_LOG=events=debug
```
or
```
RUST_LOG=events::<chai_name>=debug
```
for a particular chain

Reply:
ahaaa, cool, thank you, that `RUST_LOG=events=debug` works,
closing this :)

Reply:
Great, there are also logs for the messages.
Similar to events but with targets: `hrmp`, `ump`, `dmp` and `bridge`